roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala

```diff
@@ -32,78 +33,59 @@ object VirtualMachine {
+  val doNothing    = pure[VMState, Unit](())
+  val vmError      = modify[VMState](_.copy(exitFlag = true, exitCode = 1))
+
   def handleApplyPrimSuspend(op: Op): Unit                = ()
   def handleApplyPrimUpcall(op: Op, tag: Location): Unit  = ()
   def handleFormalsMismatch(formals: Template): Ob        = null
   def handleMissingBinding(key: Ob, argReg: Location): Ob = null
   def handleSleep(): Unit                                 = ()
   def handleXmitUpcall(op: Op, tag: Location): Unit       = ()
 
+  /*
   def handleVirtualMachineError(state: VMState): VMState =
     state.ctxt.vmError(state)._2
+   */
 
   def handlePrimResult(primResult: Result, save: Ob => VMTransition[Unit]): VMTransition[Unit] =
     primResult match {
       case Right(ob) => save(ob)
 
-      case Left(DeadThread) => modify(_.copy(doNextThreadFlag = true))
+      case Left(DeadThread) => doNextThread
 
-      case Left(PrimNotFound) => modify(_.copy(exitFlag = true, exitCode = 1))
+      case Left(PrimNotFound) => vmError
 
       // TODO: Port handleException for OTsysval
-      case Left(_) => modify(_.copy(doNextThreadFlag = true))
+      case Left(_) => doNextThread
     }
 
   def runPrim(unwind: Boolean, optPrim: Option[Prim]): VMTransition[Result] =
-    State { state =>
-      optPrim match {
-        case Some(prim) =>
-          if (unwind)
-            unwindAndApplyPrim(prim, state).swap
-          else
-            (state, prim.dispatchHelper(state.ctxt))
-
-        case None =>
-          (state, Left(PrimNotFound))
-      }
-    }
-
-  /**
-    *  This code protects the current argvec, temporarily replacing it
-    *  with the unwound argvec for use by the primitive, and then
-    *  restoring it after the primitive has finished.  This is necessary
-    *  because of the way that the compiler permits inlined primitives
-    *  (the subjects of opApplyPrim opcodes) to share a common argvec.
-    *  Unwinding cannot be permitted to clobber the argvec that the
-    *  compiler has set up, or bad things can happen (and they are *hard*
-    *  to track down).
-    */
-  def unwindAndApplyPrim(prim: Prim, state: VMState): (Result, VMState) =
-    state.ctxt.argvec.flattenRest() match {
-      case Right(newArgvec) =>
-        val tmpState = state
-          .set(_ >> 'ctxt >> 'argvec)(newArgvec)
-          .set(_ >> 'ctxt >> 'nargs)(newArgvec.elem.size)
-
-        val result = prim.dispatchHelper(tmpState.ctxt)
-
-        (result, state)
-
-      case Left(AbsentRest) =>
-        val tmpState = state
-          .set(_ >> 'ctxt >> 'argvec)(Tuple.NIL)
-          .set(_ >> 'ctxt >> 'nargs)(0)
-
-        val result = prim.dispatchHelper(tmpState.ctxt)
-
-        (result, state)
-
-      case Left(InvalidRest) =>
-        val (error, errorState) =
-          prim.runtimeError("&rest value is not a tuple", state)
+    for {
+      result <- optPrim match {
+                 case Some(prim) =>
```

Comment:
Maybe using guards in such expressions makes it a little bit clearer to read?

Reply:
Done